src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -0,0 +1,371 @@
+                new InfoBarUI(title: ServicesVSResources.Use_Keybindings_for_extensions,
+                              kind: InfoBarUI.UIKind.Button,
+                              action: OpenExtensionsHyperlink),
+                new InfoBarUI(title: ServicesVSResources.Never_show_this_again,
+                              kind: InfoBarUI.UIKind.HyperLink,
+                              action: NeverShowAgain),
+                new InfoBarUI(title: "", kind: InfoBarUI.UIKind.Close,
+                              action: InfoBarClose));
+        }
+
+        private bool IsReSharperEnabled()
+        {
+            AssertIsForeground();
+
+            // Quick exit if resharper is either uninstalled or not enabled
+            if (!_resharperExtensionEnabled)
+            {
+                return false;
+            }
+
+            if (_oleCommandTarget == null)
+            {
+                var oleServiceProvider = _serviceProvider.GetService<IOleServiceProvider>();
+                _oleCommandTarget = (IOleCommandTarget)oleServiceProvider.QueryService(VSConstants.SID_SUIHostCommandDispatcher);
+            }
+
+            var cmds = new OLECMD[1];
+            cmds[0].cmdID = SuspendId;
+            cmds[0].cmdf = 0;
+
+            ErrorHandler.ThrowOnFailure(_oleCommandTarget.QueryStatus(ReSharperCommandGroup, (uint)cmds.Length, cmds, IntPtr.Zero));
+
+            // When ReSharper is enabled, the ReSharper_Suspend command has the Enabled | Supported flags. When disabled, it has Invisible | Supported.
+            return ((OLECMDF)cmds[0].cmdf).HasFlag(OLECMDF.OLECMDF_ENABLED);
+        }
+
+        private void RestoreVsKeybindings()
+        {
+            AssertIsForeground();
+
+            if (_uiShell == null)
+            {
+                _uiShell = _serviceProvider.GetService<IVsUIShell, SVsUIShell>();
+            }
+
+            ErrorHandler.ThrowOnFailure(_uiShell.PostExecCommand(
+                    VSConstants.GUID_VSStandardCommandSet97,
+                    (uint)VSConstants.VSStd97CmdID.CustomizeKeyboard,
+                    (uint)OLECMDEXECOPT.OLECMDEXECOPT_DODEFAULT,
+                    null));
+
+            KeybindingsResetLogger.Log("KeybindingsReset");
+
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+        }
+
+        private void OpenExtensionsHyperlink()
+        {
+            ThisCanBeCalledOnAnyThread();
+            Process.Start(KeybindingsFwLink);
+
+            KeybindingsResetLogger.Log("ExtensionsLink");
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+        }
+
+        private void NeverShowAgain()
+        {
+            AssertIsForeground();
+
+            _workspace.Options = _workspace.Options.WithChangedOption(KeybindingResetOptions.NeverShowAgain, true);
+            KeybindingsResetLogger.Log("NeverShowAgain");
```

Comment:
if Exec can run from any thread, then there can be pending Task that set to run UpdateStateMachine after NeverShowAgain runs. I think this code should set NeedsReset to false as well.